File: Backend/src/Fibertest30.Api/Services/PrometheusService.cs
using Grpc.Core;
using MediatR;

namespace Fibertest30.Api;

public class PrometheusService: Prometheus.PrometheusBase
{
    private readonly ISender _mediator;

    public PrometheusService(ISender mediator)
    {
        _mediator = mediator;
    }

    public override async Task<GetCumulativeStatsResponse> GetCumulativeStats(GetCumulativeStatsRequest request, ServerCallContext context)
    {
        var timeRange = request.TimeRange;
        var monitoringPortId = request.MonitoringPortId;
        var metricName = request.MetricName;

        GetCumulativeStatsQuery query = new (timeRange, monitoringPortId, metricName);

        var metrics = await _mediator.Send(query, context.CancellationToken);

        var response = new GetCumulativeStatsResponse
        {
            Metrics = { metrics.Select(m => m.ToProto()) }
        };

        return response;
    }

    public override async Task<GetFiberSectionStatsResponse> GetFiberSectionStats(GetFiberSectionStatsRequest request, ServerCallContext context)
    {
        var timeRange = request.TimeRange;
        var monitoringPortId = request.MonitoringPortId;
        var metricName = request.MetricName;

        GetFiberSectionStatsQuery query = new (timeRange, monitoringPortId, metricName);

        var metrics = await _mediator.Send(query, context.CancellationToken);

        var response = new GetFiberSectionStatsResponse
        {
            Metrics = { metrics.Select(m => m.ToProto()) }
        };

        return response;
    }

    public override async Task<GetFiberEventStatsResponse> GetFiberEventStats(GetFiberEventStatsRequest request, ServerCallContext context)
    {
        var timeRange = request.TimeRange;
        var monitoringPortId = request.MonitoringPortId;
        var metricName = request.MetricName;

        GetFiberEventStatsQuery query = new (timeRange, monitoringPortId, metricName);

        var metrics = await _mediator.Send(query, context.CancellationToken);

        var response = new GetFiberEventStatsResponse
        {
            Metrics = { metrics.Select(m => m.ToProto()) }
        };

        return response;
    }
}

File: WebFt30/src/app/core/grpc/services/prometheus.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { GrpcWebFetchTransport } from '@protobuf-ts/grpcweb-transport';

import { AuthInterceptor } from '../auth.interceptor';
import { GrpcUtils } from '../grpc.utils';
import * as grpc from 'src/grpc-generated';

@Injectable({
  providedIn: 'root'
})
export class PrometheusService {
  private authInterceptor = inject(AuthInterceptor);
  private client: grpc.PrometheusClient;

  constructor() {
    const interceptors = [this.authInterceptor.toRpcInterceptor()];
    const options = GrpcUtils.getGrpcOptions(GrpcUtils.getApiUrl(), interceptors);
    const transport = new GrpcWebFetchTransport(options);
    this.client = new grpc.PrometheusClient(transport);
  }

  getCumulativeStats(timeRange: string, monitoringPortId: number, metricName: string): Observable<grpc.GetCumulativeStatsResponse> {
    const request: grpc.GetCumulativeStatsRequest = { metricName, timeRange, monitoringPortId };
    return GrpcUtils.unaryToObservable(
      this.client.getCumulativeStats.bind(this.client),
      request,
      {}
    );
  }

  getFiberSectionStats(timeRange: string, monitoringPortId: number, metricName: string): Observable<grpc.GetFiberSectionStatsResponse> {
    const request: grpc.GetFiberSectionStatsRequest = { metricName, timeRange, monitoringPortId };
    return GrpcUtils.unaryToObservable(
      this.client.getFiberSectionStats.bind(this.client),
      request,
      {}
    );
  }

  getFiberEventStats(timeRange: string, monitoringPortId: number, metricName: string): Observable<grpc.GetFiberEventStatsResponse> {
    const request: grpc.GetFiberEventStatsRequest = { metricName, timeRange, monitoringPortId };
    return GrpcUtils.unaryToObservable(
      this.client.getFiberEventStats.bind(this.client),
      request,
      {}
    );
  }
}
